[Bug]: CF response header policy: cannot set Server-Timing header sampling_rate to 0

### Terraform Core Version

1.3.4

### AWS Provider Version

4.38.0

### Affected Resource(s)

* aws_cloudfront_response_headers_policy

### Expected Behavior

Terraform should set the sampling rate to 0, when configured.

### Actual Behavior

During apply, we get an Amazon error (below):

### Relevant Error/Panic Output Snippet

```shell
╷
│ Error: error updating CloudFront Response Headers Policy (35ef1f9d-ff86-47df-bd10-27f041f0ea4c): InvalidArgument: The parameter STH Sampling Rate   is required.
│ 	status code: 400, request id: 85f40acd-10f2-4b74-b552-0de7679ed22b
│ 
│   with module.cf_auth.aws_cloudfront_response_headers_policy.this,
│   on ../../modules/cf_auth/main.tf line 7, in resource "aws_cloudfront_response_headers_policy" "this":
│    7: resource "aws_cloudfront_response_headers_policy" "this" {
│
```


### Terraform Configuration Files

```hcl
resource "aws_cloudfront_response_headers_policy" "this" {
  name = "foo"

  server_timing_headers_config {
    enabled       = true
    sampling_rate = 0
  }
}
```

### Steps to Reproduce

1. apply the above config

### Debug Output

```shell
module.cf_auth["auth.slido-staging.com"].aws_cloudfront_response_headers_policy.this: Modifying... [id=35ef1f9d-ff86-47df-bd10-27f041f0ea4c]
2022-11-09T13:59:33.217+0100 [INFO]  Starting apply for module.cf_auth["auth.slido-staging.com"].aws_cloudfront_response_headers_policy.this
2022-11-09T13:59:33.217+0100 [DEBUG] module.cf_auth["auth.slido-staging.com"].aws_cloudfront_response_headers_policy.this: applying the planned Update change
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: [DEBUG] Updating CloudFront Response Headers Policy: ({
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:   Id: "35ef1f9d-ff86-47df-bd10-27f041f0ea4c",
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:   IfMatch: "E23ZP02F085DFQ",
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:   ResponseHeadersPolicyConfig: {
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:     Name: "auth_slido_staging_com",
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:     SecurityHeadersConfig: {
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:       ContentSecurityPolicy: {
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         ContentSecurityPolicy: "frame-ancestors 'none'",
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         Override: false
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:       },
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:       FrameOptions: {
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         FrameOption: "DENY",
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         Override: false
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:       },
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:       StrictTransportSecurity: {
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         AccessControlMaxAgeSec: 31536000,
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         IncludeSubdomains: true,
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         Override: false,
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:         Preload: true
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:       }
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:     },
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:     ServerTimingHeadersConfig: {
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:       Enabled: true
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:     }
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5:   }
2022-11-09T13:59:33.219+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: })
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: [DEBUG] [aws-sdk-go] DEBUG: Request cloudfront/UpdateResponseHeadersPolicy2020_05_31 Details:
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: ---[ REQUEST POST-SIGN ]-----------------------------
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: PUT /2020-05-31/response-headers-policy/35ef1f9d-ff86-47df-bd10-27f041f0ea4c HTTP/1.1
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Host: cloudfront.amazonaws.com
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: User-Agent: APN/1.0 HashiCorp/1.0 Terraform/1.3.4 (+https://www.terraform.io) terraform-provider-aws/dev (+https://registry.terraform.io/providers/hashicorp/aws) aws-sdk-go/1.44.129 (go1.19.2; linux; amd64)
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Content-Length: 711
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Authorization: AWS4-HMAC-SHA256 Credential=AKIAUCPO6SDXKEGVTED3/20221109/us-east-1/cloudfront/aws4_request, SignedHeaders=content-length;host;if-match;x-amz-date, Signature=0a7c0f4fef0c2c429222535f78f704016aaca7fd8b303c43b4d77d27921ffcc2
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: If-Match: E23ZP02F085DFQ
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: X-Amz-Date: 20221109T125933Z
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Accept-Encoding: gzip
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: 
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: <ResponseHeadersPolicyConfig xmlns="http://cloudfront.amazonaws.com/doc/2020-05-31/"><Name>auth_slido_staging_com</Name><SecurityHeadersConfig><ContentSecurityPolicy><Override>false</Override><ContentSecurityPolicy>frame-ancestors &#39;none&#39;</ContentSecurityPolicy></ContentSecurityPolicy><FrameOptions><FrameOption>DENY</FrameOption><Override>false</Override></FrameOptions><StrictTransportSecurity><AccessControlMaxAgeSec>31536000</AccessControlMaxAgeSec><IncludeSubdomains>true</IncludeSubdomains><Override>false</Override><Preload>true</Preload></StrictTransportSecurity></SecurityHeadersConfig><ServerTimingHeadersConfig><Enabled>true</Enabled></ServerTimingHeadersConfig></ResponseHeadersPolicyConfig>
2022-11-09T13:59:33.220+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: -----------------------------------------------------
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: [DEBUG] [aws-sdk-go] DEBUG: Response cloudfront/UpdateResponseHeadersPolicy2020_05_31 Details:
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: ---[ RESPONSE ]--------------------------------------
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: HTTP/1.1 400 Bad Request
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Connection: close
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Content-Length: 295
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Content-Type: text/xml
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: Date: Wed, 09 Nov 2022 12:59:32 GMT
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: X-Amzn-Requestid: 1cb99755-75a1-4fd2-adc5-041a4f4513ee
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: 
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: 
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: -----------------------------------------------------
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: [DEBUG] [aws-sdk-go] <?xml version="1.0"?>
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: <ErrorResponse xmlns="http://cloudfront.amazonaws.com/doc/2020-05-31/"><Error><Type>Sender</Type><Code>InvalidArgument</Code><Message>The parameter STH Sampling Rate   is required.</Message></Error><RequestId>1cb99755-75a1-4fd2-adc5-041a4f4513ee</RequestId></ErrorResponse>
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: [DEBUG] [aws-sdk-go] DEBUG: Validate Response cloudfront/UpdateResponseHeadersPolicy2020_05_31 failed, attempt 0/25, error InvalidArgument: The parameter STH Sampling Rate   is required.
2022-11-09T13:59:33.629+0100 [DEBUG] provider.terraform-provider-aws_v4.38.0_x5: 	status code: 400, request id: 1cb99755-75a1-4fd2-adc5-041a4f4513ee
2022-11-09T13:59:33.629+0100 [ERROR] provider.terraform-provider-aws_v4.38.0_x5: Response contains error diagnostic: tf_provider_addr=registry.terraform.io/hashicorp/aws @caller=github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/diag/diagnostics.go:55 diagnostic_detail= diagnostic_severity=ERROR diagnostic_summary="error updating CloudFront Response Headers Policy (35ef1f9d-ff86-47df-bd10-27f041f0ea4c): InvalidArgument: The parameter STH Sampling Rate   is required.
	status code: 400, request id: 1cb99755-75a1-4fd2-adc5-041a4f4513ee" tf_proto_version=5.3 tf_rpc=ApplyResourceChange @module=sdk.proto tf_req_id=b3bc4e8e-584d-8d88-5fa7-427749c77bbe tf_resource_type=aws_cloudfront_response_headers_policy timestamp=2022-11-09T13:59:33.629+0100
2022-11-09T13:59:33.631+0100 [ERROR] vertex "module.cf_auth[\"auth.slido-staging.com\"].aws_cloudfront_response_headers_policy.this" error: error updating CloudFront Response Headers Policy (35ef1f9d-ff86-47df-bd10-27f041f0ea4c): InvalidArgument: The parameter STH Sampling Rate   is required.
	status code: 400, request id: 1cb99755-75a1-4fd2-adc5-041a4f4513ee
```

### Panic Output

_No response_

### Important Factoids

`sample_rate = 0` is a valid case. You enabled the "Server-Timing header" feature, but don't send the timing response header for any request by default. If you would like to see the server timing, you can then do it per-request by providing "Pragma: server-timing" request header.

### References

https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/understanding-response-headers-policies.html#server-timing-header-sampling-rate

### Would you like to implement a fix?

No

--- a/.changelog/27778.txt
+++ b/.changelog/27778.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_cloudfront_response_headers_policy: Allow `server_timing_headers_config.0.sampling_rate` to be `0`
+```

--- a/internal/service/cloudfront/response_headers_policy.go
+++ b/internal/service/cloudfront/response_headers_policy.go
@@ -1114,7 +1114,7 @@ func expandResponseHeadersPolicyServerTimingHeadersConfig(tfMap map[string]inter
 		apiObject.Enabled = aws.Bool(v)
 	}
 
-	if v, ok := tfMap["sampling_rate"].(float64); ok && v != 0 {
+	if v, ok := tfMap["sampling_rate"].(float64); ok {
 		apiObject.SamplingRate = aws.Float64(v)
 	}
 

--- a/internal/service/cloudfront/response_headers_policy_test.go
+++ b/internal/service/cloudfront/response_headers_policy_test.go
@@ -269,7 +269,7 @@ func TestAccCloudFrontResponseHeadersPolicy_ServerTimingHeadersConfig(t *testing
 			{
 				Config: testAccResponseHeadersPolicyConfig_serverTiming(rName, true, 0),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckResponseHeadersPolicyExists(resourceName),
+					testAccCheckResponseHeadersPolicyExists(ctx, resourceName),
 					resource.TestCheckResourceAttr(resourceName, "comment", ""),
 					resource.TestCheckResourceAttr(resourceName, "cors_config.#", "0"),
 					resource.TestCheckResourceAttr(resourceName, "custom_headers_config.#", "0"),

--- a/internal/service/cloudfront/response_headers_policy_test.go
+++ b/internal/service/cloudfront/response_headers_policy_test.go
@@ -15,7 +15,7 @@ import (
 	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
 )
 
-func TestAccCloudFrontResponseHeadersPolicy_CorsConfig(t *testing.T) {
+func TestAccCloudFrontResponseHeadersPolicy_cors(t *testing.T) {
 	ctx := acctest.Context(t)
 	rName1 := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	rName2 := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
@@ -94,7 +94,7 @@ func TestAccCloudFrontResponseHeadersPolicy_CorsConfig(t *testing.T) {
 	})
 }
 
-func TestAccCloudFrontResponseHeadersPolicy_CustomHeadersConfig(t *testing.T) {
+func TestAccCloudFrontResponseHeadersPolicy_customHeaders(t *testing.T) {
 	ctx := acctest.Context(t)
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_cloudfront_response_headers_policy.test"
@@ -139,7 +139,7 @@ func TestAccCloudFrontResponseHeadersPolicy_CustomHeadersConfig(t *testing.T) {
 	})
 }
 
-func TestAccCloudFrontResponseHeadersPolicy_SecurityHeadersConfig(t *testing.T) {
+func TestAccCloudFrontResponseHeadersPolicy_securityHeaders(t *testing.T) {
 	ctx := acctest.Context(t)
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_cloudfront_response_headers_policy.test"
@@ -219,7 +219,7 @@ func TestAccCloudFrontResponseHeadersPolicy_SecurityHeadersConfig(t *testing.T) 
 	})
 }
 
-func TestAccCloudFrontResponseHeadersPolicy_ServerTimingHeadersConfig(t *testing.T) {
+func TestAccCloudFrontResponseHeadersPolicy_serverTimingHeaders(t *testing.T) {
 	ctx := acctest.Context(t)
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_cloudfront_response_headers_policy.test"

--- a/internal/service/cloudfront/response_headers_policy_test.go
+++ b/internal/service/cloudfront/response_headers_policy_test.go
@@ -267,6 +267,21 @@ func TestAccCloudFrontResponseHeadersPolicy_ServerTimingHeadersConfig(t *testing
 				),
 			},
 			{
+				Config: testAccResponseHeadersPolicyConfig_serverTiming(rName, true, 0),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckResponseHeadersPolicyExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "comment", ""),
+					resource.TestCheckResourceAttr(resourceName, "cors_config.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "custom_headers_config.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "security_headers_config.#", "0"),
+					resource.TestCheckResourceAttrSet(resourceName, "etag"),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
+					resource.TestCheckResourceAttr(resourceName, "server_timing_headers_config.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "server_timing_headers_config.0.enabled", "true"),
+					resource.TestCheckResourceAttr(resourceName, "server_timing_headers_config.0.sampling_rate", "0"),
+				),
+			},
+			{
 				Config: testAccResponseHeadersPolicyConfig_serverTiming(rName, false, 0),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckResponseHeadersPolicyExists(ctx, resourceName),

